### ISSUE TITLE
Possible Bug on Syntax Highlight or code formatting under (La)TeX Option

### ISSUE THREAD

**Issue by T-AbX** [texstudio-org/texstudio#3684]
I am experiencing an issue with TeXstudio version 4.8.1 that did not occur in version 4.1.2. This issue might be related to syntax highlighting or code formatting under Visualize > Highlight > (La)TeX.

I will try to explain with a very simple example. Loading the amsmath and xcolor packages (\usepackage{amsmath} and \usepackage{xcolor}) and creating a simple equation like this:

**\begin{equation}
    \boxed{
    \begin{aligned}
        AAA
    \end{aligned}
}
\end{equation}
AAAAA**

**\begin{equation}
    \boxed{
    \begin{aligned}
        BBBB
    \end{aligned}
}
\end{equation}
BBBB**

The second equation is mysteriously misinterpreted by TeXstudio, which points to an incorrect error at \end{equation}, reporting that an environment is being closed that was not opened. Therefore, all the content of the file that follows \end{equation} is shown in green, as if the equation were still open. Indeed, this is incorrect, as can be seen in the simple code and it compiles very easily without errors, as can be tested. But TeXstudio is detecting an error that should not exist in this simple example, as well as in others.

After checking many TeXstudio options and making several attempts, I realized that none of my actions could change this behavior. It is indeed strange. This does not happen in version 4.1.2.

Could this be a bug? Or is there something else wrong?

I will send some images of the exposed code and others to help with understanding.

I have tried uninstalling and reinstalling cleanly several times, including doing the same with MikTeX, but the problem persists (Windows 10 x64). The same happened in other tests on Debian, Red Hat, and Slackware systems.

I greatly appreciate the developers' attention. TeXstudio has always been my favorite editor, but this issue has been bothering me a lot recently.

<img width="181" alt="1" src="https://github.com/texstudio-org/texstudio/assets/172121853/739c3c56-5302-44f2-b57e-b1a546a28b93">
<img width="796" alt="2" src="https://github.com/texstudio-org/texstudio/assets/172121853/a6dd9f50-9ede-46b9-a063-fe1012ebcc18">

P.S. (more information): Also (sorry for not mentioning this before, hence this edit), the problem seems to be very related to indentation, although the environments are opened and closed correctly, as can be easily tested. However, even when testing with the same indentation, the issue occurs depending on the order in which the text is written. For example, if I simply write the text step by step, the bug occurs. If I cut from \begin{aligned} to \end{aligned} and paste it inside \boxed{}, TeXstudio does not point out the error.

**Comment by sunderme** [comment]
so, I have tried your example and it does not give an issue.
<img width="349" alt="grafik" src="https://github.com/texstudio-org/texstudio/assets/14033169/9233e73e-19d1-4738-a314-bb9e78af4411">
So, my guess is that you have not checked that your example stand-alone causes the issue ?

That said, looking at the rest of the screenshot, it seems that the content of `\box{...}` spreads over several lines. txs has a "runaway" limit which, by default, is 30 lines. Longer arguments are ignored and obviously the parsing can get messed up.

This can be changed by the user by manually editing in texstudio.ini the line `Editor\RUNAWAYLIMIT=30` and set the value to something larger (100?)

**Comment by sunderme** [comment]
see also https://github.com/texstudio-org/texstudio/wiki/Frequently-Asked-Questions#where-are-the-settings-stored

**Comment by T-AbX** [comment]
First of all, thank you very much for the help and the quick and efficient response, sunderme! **It solved my problem!**

_Additional comments on the test:_

_I conducted several tests in various files to arrive at a simple model to identify the error, to check if it was not caused by my specific example. This is why in the Bug Report request I presented an example with a very simple equation. To verify the possibility of the error being due to some mistake I might be making, I uninstalled and performed a clean installation on all systems several times. But the error always occurred._

_Based on these tests, I concluded (without ruling out the possibility of being wrong) that TeXstudio was identifying the "error" mainly by interpreting the indentation, even with the {aligned} and {boxed} environments correctly opened and closed. Curiously, writing step by step the simple equation of the presented example, TeXstudio understood and pointed out the error message, when hovering the mouse over \end{equation}, that it was "closing environment that was not opened". Rewriting the same equation by cutting the part between \begin{aligned} and \end{aligned}, and pasting it inside \boxed{} made TeXstudio understand correctly, not pointing out the error. In this sense, I was, in some way, managing to bypass or force TeXstudio's understanding. However, this was unnecessarily complicated because, although I have the habit of following indentations in minute details, in long tests it was sometimes easier and faster to simply include a boxed directly in an important equation, and thus the error appeared._

Without further ado, **you are absolutely correct!** I followed your invaluable instructions, closed TeXstudio, edited the texstudio.ini file in the RUNAWAYLIMIT section to Editor\RUNAWAYLIMIT=100 on all tested operating systems (Windows 10, Debian, Red Hat, and Slackware). I reopened it and all previously pointed out errors in TeXstudio, both in the simple tested examples and in others with extremely long equations, were resolved immediately and efficiently!

I was curious that the same error did not occur with version 4.1.2 with the Editor\RUNAWAYLIMIT=30 configuration. Also, the error occurred in version 4.8.1 (with the same RUNAWAYLIMIT configuration) with equations where the content described inside {aligned} had 1 or 2 lines. It doesn't matter! **Increasing the limit value solves the problem!** Thank you very much! Would it be interesting to include a higher default value in future updates?

I have nothing but thanks! Thank you very much! TeXstudio has always been my favorite editor by far, much better than all the others I have tested over decades, much better than Texmaker, Overleaf, among others. And now I am certain and convinced that it is the best because of all your professional and helpful work! I will make a donation! Thank you!!
